pkgs/by-name/wa/warp-terminal/update.sh

```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env nix-shell
+#!nix-shell -I nixpkgs=./. -i bash -p curl jq common-updater-scripts nix-prefetch moreutils
+#shellcheck shell=bash
+set -eu -o pipefail
+
+dirname="$(dirname "$0")"
+
+json_get() {
+    jq -r "$1" < "$dirname/versions.json"
+}
+
+json_set() {
+    jq --arg x "$2" "$1 = \$x" < "$dirname/versions.json" | sponge "$dirname/versions.json"
+}
+
+# it's fine to use just the linux URL because the package also assumes the versioning is the same between darwin and linux
+version=$(curl -Ls -o /dev/null -w %{url_effective} https://app.warp.dev/download\?package\=pacman 2>&1 | grep -oP -m 1 "(?<=v)[\d.\w]+" | head -1)
```

Comment:
I decided to keep it in because it's writing the same version to both, though if it's confusing I could remove it. And I tried not pulling the entire file, but otherwise it wouldn't get the final destination url. the `-I` flag is what that's supposed to do but it doesn't follow through all the redirects.

Reply:
> I decided to keep it in because it's writing the same version to both, though if it's confusing I could remove it. And I tried not pulling the entire file, but otherwise it wouldn't get the final destination url. the `-I` flag is what that's supposed to do but it doesn't follow through all the redirects.

there is only one redirect. not all servers support HEAD, which is what -I is doing, and this one doesn't.

having multiple variables containing the same value is sloppy. granted in this case it is pretty minor.

Reply:
here is a function that will resolve the url. started with `http` rather than `https` to get 2 redirects.

```bash
#!/usr/bin/env bash

set -e

err() {
    echo "$*" >&2
}

resolve_url() {
    local url="$1"
    local -i i max_retries
    ((max_retries = 30))
    url="$1"
    for ((i = 0; i < max_retries; i++)); do
        echo "round $i"
        url="$(curl -s -o /dev/null -w %{redirect_url} "${url}")"
        [[ "$url" != *stable/v* ]] || break
    done
    ((i < max_retries)) || { err "too many retries"; exit 1; }
    echo $url
}

resolve_url "http://app.warp.dev/download?package=pacman"
```

Reply:
Wow, thanks! I didn't even know about servers not supporting HEAD. And yeah, I don't know why I didn't wrap the JSON stuff in a loop.